File: backend/app/services/media.py
from __future__ import annotations

import asyncio
import json
import os
import re
from datetime import UTC, datetime
from pathlib import Path
from uuid import UUID, uuid4

from fastapi import UploadFile

from app.config import Settings
from app.models import MediaMetadata, MediaResponse
from app.services.registry import MediaRegistry


class InvalidMediaError(ValueError):
    pass


class UploadTooLargeError(ValueError):
    pass


def _fps(value: str | None) -> float:
    if not value:
        return 0
    numerator, _, denominator = value.partition("/")
    try:
        return float(numerator) / float(denominator or 1)
    except (ValueError, ZeroDivisionError):
        return 0


async def _command(*args: str, timeout: float = 60) -> tuple[bytes, bytes]:
    process = await asyncio.create_subprocess_exec(
        *args,
        stdout=asyncio.subprocess.PIPE,
        stderr=asyncio.subprocess.PIPE,
    )
    try:
        stdout, stderr = await asyncio.wait_for(process.communicate(), timeout)
    except TimeoutError:
        process.kill()
        await process.wait()
        raise InvalidMediaError("media inspection timed out")
    if process.returncode:
        message = stderr.decode(errors="replace")[-1000:].strip()
        raise InvalidMediaError(message or f"{args[0]} failed")
    return stdout, stderr


async def probe_media(ffprobe: str, path: Path) -> dict[str, object]:
    stdout, _ = await _command(
        ffprobe,
        "-v",
        "error",
        "-print_format",
        "json",
        "-show_format",
        "-show_streams",
        str(path),
    )
    try:
        data = json.loads(stdout)
    except json.JSONDecodeError as exc:
        raise InvalidMediaError("ffprobe returned invalid JSON") from exc
    video = next(
        (stream for stream in data.get("streams", []) if stream.get("codec_type") == "video"),
        None,
    )
    audio = next(
        (stream for stream in data.get("streams", []) if stream.get("codec_type") == "audio"),
        None,
    )
    if not video or not video.get("width") or not video.get("height"):
        raise InvalidMediaError("uploaded file has no valid video stream")
    duration = float(data.get("format", {}).get("duration") or video.get("duration") or 0)
    frame_rate = _fps(video.get("avg_frame_rate") or video.get("r_frame_rate"))
    if duration <= 0 or frame_rate <= 0:
        raise InvalidMediaError("video duration or frame rate is invalid")
    return {
        "duration": duration,
        "width": int(video["width"]),
        "height": int(video["height"]),
        "fps": frame_rate,
        "has_audio": audio is not None,
        "codec": video.get("codec_name"),
    }


async def generate_thumbnail(ffmpeg: str, source: Path, destination: Path, duration: float) -> None:
    temporary = destination.with_suffix(".tmp.jpg")
    temporary.unlink(missing_ok=True)
    try:
        await _command(
            ffmpeg,
            "-v",
            "error",
            "-ss",
            str(min(max(duration * 0.1, 0), 5)),
            "-i",
            str(source),
            "-frames:v",
            "1",
            "-vf",
            "scale=480:-2",
            "-y",
            str(temporary),
            timeout=120,
        )
        os.replace(temporary, destination)
    finally:
        temporary.unlink(missing_ok=True)


def _safe_extension(filename: str | None) -> str:
    suffix = Path(filename or "").suffix.lower()
    return suffix if re.fullmatch(r"\.[a-z0-9]{1,8}", suffix) else ".bin"


async def ingest_upload(
    upload: UploadFile,
    settings: Settings,
    registry: MediaRegistry,
) -> MediaResponse:
    media_id = uuid4()
    incoming = registry.media_dir / f".{media_id}.upload"
    final = registry.media_dir / f"{media_id}{_safe_extension(upload.filename)}"
    size = 0
    try:
        with incoming.open("xb") as stream:
            while chunk := await upload.read(1024 * 1024):
                size += len(chunk)
                if size > settings.max_upload_bytes:
                    raise UploadTooLargeError(
                        f"upload exceeds {settings.max_upload_bytes} byte limit"
                    )
                stream.write(chunk)
            stream.flush()
            os.fsync(stream.fileno())
        if size == 0:
            raise InvalidMediaError("uploaded file is empty")
        probed = await probe_media(settings.ffprobe, incoming)
        os.replace(incoming, final)
        thumbnail = registry.thumbnail_path(media_id)
        await generate_thumbnail(
            settings.ffmpeg, final, thumbnail, float(probed["duration"])
        )
        metadata = MediaMetadata(
            fileId=media_id,
            original_name=Path(upload.filename or "video").name,
            storage_name=final.name,
            content_type=upload.content_type or "application/octet-stream",
            size=size,
            created_at=datetime.now(UTC),
            **probed,
        )
        registry.save(metadata)
        return media_response(metadata)
    except Exception:
        incoming.unlink(missing_ok=True)
        final.unlink(missing_ok=True)
        registry.thumbnail_path(media_id).unlink(missing_ok=True)
        registry.metadata_path(media_id).unlink(missing_ok=True)
        raise
    finally:
        await upload.close()


def media_response(metadata: MediaMetadata) -> MediaResponse:
    media_id = metadata.media_id
    return MediaResponse(
        fileId=media_id,
        filename=str(media_id),
        original_name=metadata.original_name,
        url=f"/api/media/{media_id}/stream",
        thumbnail_url=f"/api/media/{media_id}/thumbnail",
        duration=metadata.duration,
        width=metadata.width,
        height=metadata.height,
        fps=metadata.fps,
        has_audio=metadata.has_audio,
        codec=metadata.codec,
    )

File: server/services/media.js
const { execFile, spawn } = require('child_process');
const path = require('path');
const fs = require('fs');
const { v4: uuidv4 } = require('uuid');

function runCommand(cmd, args, timeoutMs = 60000) {
  return new Promise((resolve, reject) => {
    execFile(cmd, args, { timeout: timeoutMs, maxBuffer: 10 * 1024 * 1024 }, (err, stdout, stderr) => {
      if (err) {
        return reject(new Error(stderr || err.message));
      }
      resolve({ stdout, stderr });
    });
  });
}

function parseFps(str) {
  if (!str) return 0;
  const parts = str.split('/');
  if (parts.length === 2) {
    const num = parseFloat(parts[0]);
    const den = parseFloat(parts[1]) || 1;
    return num / den;
  }
  return parseFloat(str) || 0;
}

const IMAGE_CODECS = new Set(['mjpeg', 'jpeg2000', 'png', 'gif', 'webp', 'bmp', 'tiff', 'heif', 'heic']);

async function probeMedia(ffprobeBin, mediaPath) {
  const { stdout } = await runCommand(ffprobeBin, [
    '-v', 'error',
    '-print_format', 'json',
    '-show_format',
    '-show_streams',
    mediaPath
  ]);

  let data;
  try {
    data = JSON.parse(stdout);
  } catch (e) {
    throw new Error('Invalid ffprobe output');
  }

  const streams = data.streams || [];
  const video = streams.find(s => s.codec_type === 'video');
  const audio = streams.find(s => s.codec_type === 'audio');

  if (!video || !video.width || !video.height) {
    throw new Error('Uploaded file has no valid video stream');
  }

  // 이미지 파일 판별 (duration/fps 없음)
  const isImage = IMAGE_CODECS.has(video.codec_name) ||
    video.nb_frames === '1' ||
    (video.r_frame_rate === '25/1' && !data.format?.duration && video.codec_name !== 'h264');

  if (isImage) {
    return {
      duration: 5,  // 기본 5초로 타임라인에 배치
      width: parseInt(video.width, 10),
      height: parseInt(video.height, 10),
      fps: 1,
      has_audio: false,
      codec: video.codec_name || 'mjpeg',
      is_image: true
    };
  }

  const duration = parseFloat(data.format?.duration || video.duration || 0);
  const fps = parseFps(video.avg_frame_rate || video.r_frame_rate);

  if (duration <= 0 || fps <= 0) {
    throw new Error('Invalid video duration or frame rate');
  }

  return {
    duration,
    width: parseInt(video.width, 10),
    height: parseInt(video.height, 10),
    fps,
    has_audio: Boolean(audio),
    codec: video.codec_name || 'h264',
    is_image: false
  };
}

async function generateThumbnail(ffmpegBin, sourcePath, destPath, duration, isImage = false) {
  const args = isImage
    ? ['-v', 'error', '-i', sourcePath, '-frames:v', '1', '-vf', 'scale=480:-2', '-y', destPath]
    : ['-v', 'error', '-ss', Math.min(Math.max(duration * 0.1, 0), 5).toFixed(2), '-i', sourcePath, '-frames:v', '1', '-vf', 'scale=480:-2', '-y', destPath];
  await runCommand(ffmpegBin, args, 120000);
}

class MediaRegistry {
  constructor(settings) {
    this.settings = settings;
  }

  getMetadataPath(mediaId) {
    return path.join(this.settings.metadataDir, `${mediaId}.json`);
  }

  getThumbnailPath(mediaId) {
    return path.join(this.settings.thumbnailDir, `${mediaId}.jpg`);
  }

  saveMetadata(metadata) {
    const filePath = this.getMetadataPath(metadata.media_id);
    fs.writeFileSync(filePath, JSON.stringify(metadata, null, 2), 'utf-8');
  }

  getMetadata(mediaId) {
    const filePath = this.getMetadataPath(mediaId);
    if (!fs.existsSync(filePath)) {
      throw new Error(`Media not found: ${mediaId}`);
    }
    const raw = fs.readFileSync(filePath, 'utf-8');
    return JSON.parse(raw);
  }

  getMediaPath(metadata) {
    return path.join(this.settings.uploadDir, metadata.storage_name);
  }
}

async function ingestUpload(file, settings, registry) {
  const mediaId = uuidv4();
  const ext = path.extname(file.originalname || '').toLowerCase() || '.mp4';
  const storageName = `${mediaId}${ext}`;
  const finalPath = path.join(settings.uploadDir, storageName);

  fs.renameSync(file.path, finalPath);

  try {
    const probed = await probeMedia(settings.ffprobeBin, finalPath);
    const thumbnailPath = registry.getThumbnailPath(mediaId);
    await generateThumbnail(settings.ffmpegBin, finalPath, thumbnailPath, probed.duration, probed.is_image);

    const metadata = {
      media_id: mediaId,
      fileId: mediaId,
      original_name: file.originalname || 'video',
      storage_name: storageName,
      content_type: file.mimetype || 'video/mp4',
      size: file.size,
      created_at: new Date().toISOString(),
      ...probed
    };

    registry.saveMetadata(metadata);

    return {
      fileId: mediaId,
      media_id: mediaId,
      filename: mediaId,
      original_name: metadata.original_name,
      url: `/api/media/${mediaId}/stream`,
      thumbnail_url: `/api/media/${mediaId}/thumbnail`,
      duration: metadata.duration,
      width: metadata.width,
      height: metadata.height,
      fps: metadata.fps,
      has_audio: metadata.has_audio,
      codec: metadata.codec,
      is_image: metadata.is_image || false
    };
  } catch (err) {
    if (fs.existsSync(finalPath)) fs.unlinkSync(finalPath);
    const thumb = registry.getThumbnailPath(mediaId);
    if (fs.existsSync(thumb)) fs.unlinkSync(thumb);
    throw err;
  }
}

module.exports = { MediaRegistry, ingestUpload, probeMedia, generateThumbnail };
